fiftyone/core/collections.py

```diff
@@ -170,37 +176,65 @@ def save(self, sample):
 
         if self._batching_strategy == "static":
             self._curr_batch_size += 1
-            if self._curr_batch_size >= self.batch_size:
+            if (self._curr_batch_size >= self.batch_size) or (
+                self._max_batch_size
+                and self._curr_batch_size >= self._max_batch_size
+            ):
                 self._save_batch()
                 self._curr_batch_size = 0
         elif self._batching_strategy == "size":
             if sample_ops:
+                self._curr_batch_size += len(sample_ops)
                 self._curr_batch_size_bytes += sum(
                     len(str(op)) for op in sample_ops
                 )
 
             if frame_ops:
+                self._curr_batch_size += len(sample_ops)
```

Comment:
So theoretically it could happen that `frame_ops` contains 99k ops, and then after pulling in the frame ops for the next sample, it could *exceed* 100k ops (eg the video could have more than 1k frames and the user edited all of them).

Despite the promise made by the new `max_batch_size` parameter, the proposed implementation wouldn't actually prevent the "100k limit" error.

Here's an alternative strategy: *don't* add `max_batch_size=` to `SaveContext` at all. Instead, just update `bulk_write()` to this:

```py
    if batcher is False:
        batcher = partial(StaticBatcher, batch_size=-1, max_batch_size=100000)
```

Then it would gracefully split into multiple batches inside `bulk_write()` *only if necessary*.

Reply:
Go ahead and hard-code `max_batch_size=100000` here too:
https://github.com/voxel51/fiftyone/blob/7117db9e2e00688adc5b7804fe9aad84e7e95b85/fiftyone/core/utils.py#L1812

The other default batchers already encode the 100k limit.

(which means that, while batchers *could* be arbitrary things, we are still designing for the typical case that they will be used to generate bulk writes).